[bug 1279606] Move Page ID dataLayer push before GTM loads

JamesDLorence opened the pull request:
## Description
Move page ID dataLayer push to before the GTM container instantiates.

## Bugzilla link
https://bugzilla.mozilla.org/show_bug.cgi?id=1279606

JamesDLorence commented:
Alex, have you had a chance to review this yet? I have a Call with Pmac and Gareth about where we're at with making GTM CSP compliant and this is blocking me.

alexgibson commented:
> Alex, have you had a chance to review this yet? I have a Call with Pmac and Gareth about where we're at with making GTM CSP compliant and this is blocking me.

Nope sorry I have other priorities I need to take care of first. Gareth understands this work is not in our sprint planning, so I'm afraid this will be blocked until I have time to take a look.

JamesDLorence commented:
Completely understand. I just wanted to be able to communicate with them where we're at. Thanks for the update.

alexgibson commented:
No worries, really I'd like to get #4182 merged and into production first before this one (this will also likely need rebasing come then too).

alexgibson commented:
@JamesDLorence - now that #4182 has merged, can you please rebase this branch against master? Thanks

JamesDLorence commented:
Rebased and ready! 👍

alexgibson commented on media/js/base/core-datalayer.js:
```diff
@@ -2,50 +2,34 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-// create namespace
-if (typeof Mozilla == 'undefined') {
```
Because this is relying & adding to an existing namespace that's now in a different bundle altogether, I think we should probably keep this check, and also add a fallback for if `Mozilla.Analytics` is undefined also?

alexgibson commented on media/js/base/datalayer-page-id.js:
```diff
@@ -0,0 +1,32 @@
+}
+
+(function() {
+    var dataLayer = window.dataLayer = window.dataLayer || [];
```
As we're now moving this out of the GTM snippet and into an external script, it would be worthwhile adding a comment here that states this is where the core `window.dataLayer` object is now defined?

alexgibson commented:
Now that `getPageId` is moving to it's own file and has it's own test suite, should we remove the [existing duplicate test](https://github.com/mozilla/bedrock/blob/master/tests/unit/spec/base/core-datalayer.js#L10-L26)?

alexgibson commented on media/js/base/datalayer-page-id.js:
```diff
@@ -0,0 +1,32 @@
+        var pageId = document.getElementsByTagName('html')[0].getAttribute('data-gtm-page-id');
+        var pathName = path ? path : document.location.pathname;
+
+        return pageId ? pageId : pathName.replace(/\/[^\/]*/, '');
```
It would be nice to match the locale a little more explicitly here, like we did in `updateDataLayerPush`

alexgibson replied:
Once you do this, adding an extra test to make sure a URL path that does not contain a locale functions as expected would be nice too.

alexgibson commented:
Small nit: for consistency sake I think we should rename `datalayer-page-id.js` to `core-datalayer-page-id.js`

alexgibson commented:
@JamesDLorence - this is now up on demo4 for testing: https://www-demo4.allizom.org/en-US/

JamesDLorence commented:
There was an issue with the first forward slash being removed from the `pageId` when removing the locale which I fixed, and I updated the name from `datalayer-page-id.js` to `core-datalayer-page-id.js`.

alexgibson commented:
Updated demo4 with the latest code mentioned above.

JamesDLorence commented:
Everything looks good to me. Thumbs up to merge from me.

alexgibson commented:
Sorry for the delay on getting back to this, I've been occupied by other priorities this week and I need to try and resolve an issue with one of our demo servers before I take a look at manually testing this.

alexgibson commented:
I went through some testing of a few edge case scenarios and there's one recommendation I'd like to make:

If for whatever reason a page exists where the `core-datalayer-page-id.js` is not present in the head, then any call to `window.dataLayer` in the [common base bundle](https://github.com/mozilla/bedrock/blob/master/bedrock/settings/static_media.py#L1000-L1014) will now throw an error. I don't think this is a problem in reality, but I'd like to err on the side of caution and make sure a definition of `window.dataLayer` also exists in `core-datalayer-init.js` [here]https://github.com/mozilla/bedrock/blob/master/media/js/base/core-datalayer-init.js#L9). (e.g. `var dataLayer = window.dataLayer || [];`).

Aside from this one nit, this looks good to me r+wc

JamesDLorence commented:
Defined `window.dataLayer` in `core-datalayer-init.js` and rebased against master. It should be ready. 👍

alexgibson commented:
Thanks @JamesDLorence r+
